### PR TITLE
Avoid many error messages for copybook downloading

### LINKED ISSUE #384: Simplify UX for handle missing copybooks
## Description ##
As a USER, I would like to have a more consistent experience when I am dealing with missing copybooks.
Since the Cobol Language Support handles references for copybooks into the ```settings.json```
I don't need to interact with popup messages and commands because I know where I need to go to update the copybook list I'd like to use for my COBOL program.

## Acceptance Criteria ##
TBD

## Definition of Done ##
- The story is implemented according to the description.
- Unit testing done (Preferably by TDD approach, coverage should be at least 70%).
- Code review done by at least one other team member.
- Manual testing is done.
- Documentation is added and reviewed (if needed).
- The pull request is approved.
- The story is merged into the development branch.
- Tests are performed on the development branch.

## Technical details ##
This is the place for any implementation-related notes

### LINKED ISSUE #390: Simplify the UX of info message
**Is your feature request related to a problem?**

As a user I want to see just one info message, that will inform me about the absence some of copybooks and actions she may do in order to resolve them. The message should be generic and short, just to inform Michelle that some of her copybooks are not resolved and navigate her into the settings file to provide data to resolve them.

Acceptance criterias

Only ONE message regarding the missing copybooks can appear at a time!

**Describe the solution you'd like**

**Describe alternatives you've considered**

**Additional information**

### PR DESCRIPTION
**Pull request by grianbrcom**
The middleware process two calls: resolve URI and download copybook.
The server synchronously resolves URI copybooks, not in the server cache
during document analysis. After the scan, the server sends one call to
middleware with a list of copybooks to download.
DoneAnalysisEvent was added to the databus.

Closes #384
Closes #390

Signed-off-by: Anton Grigorev <anton.grigorev@broadcom.com>

### CODE REVIEW

**Comment by zacanbrcom** [comment]
+ @asatklichov

File: clients/cobol-lsp-vscode-extension/src/__tests__/MiddlewareTest.ts [hunk 1]
```diff
@@ -28,20 +28,33 @@ describe("Copybook downloader", () => {
...
         expect(await middleware.handleConfigurationRequest(params, null, null)).toEqual(["copybookUri"]);
         expect(copybookResolverURI.resolveCopybookURI).toHaveBeenCalledWith("bookName", "cobFile");
     });
     it("Handle copybook request for name with dots", async () => {
         const copybookResolverURI: any = {
             resolveCopybookURI: jest.fn().mockResolvedValue("copybookUri"),
         };
-        const middleware = new Middleware(null, copybookResolverURI);
-        const params = constructParams("broadcom-cobol-lsp.copybook.USER.CLIST.COB.bookName");
+        const middleware = new Middleware(copybookResolverURI, null);
```
Comment (asatklichov) [inline clients/cobol-lsp-vscode-extension/src/__tests__/MiddlewareTest.ts:40]: can we extract this middleware, params, etc. in desc. with let to reuse

File: clients/cobol-lsp-vscode-extension/src/services/Middleware.ts [hunk 2]
```diff
@@ -25,21 +25,28 @@ export class Middleware {
...
+            if (sectionName.startsWith("broadcom-cobol-lsp.copybook-resolve")) {
                 const [cobolFileName, copybookName] = Middleware.extractFileAndCopybookNames(sectionName);
                 return [await this.copybookResolverURI.resolveCopybookURI(copybookName, cobolFileName)];
             }
+            if (sectionName.startsWith("broadcom-cobol-lsp.copybook-download")) {
+                for (const item of params.items) {
+                    const [cobolFileName, copybookName] = Middleware.extractFileAndCopybookNames(item.section);
+                    this.copybookDownloader.downloadCopybook(cobolFileName, copybookName);
+                }
+                return [];
```
Comment (asatklichov) [inline clients/cobol-lsp-vscode-extension/src/services/Middleware.ts:48]: Regarding to this, also better to define return type in method declaration explicitly

File: com.ca.lsp.cobol/lsp-core-domain/src/main/java/com/broadcom/lsp/domain/cobol/event/model/DoneAnalysisEvent.java [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+/**
+ * This class is a data transfer object that contains the uri of analyzed document.
+ */
+@NoArgsConstructor
+@Data
+public class DoneAnalysisEvent extends DataEvent{
```
[the remark this answers is not in the document]
Comment (grianbrcom) [reply com.ca.lsp.cobol/lsp-core-domain/src/main/java/com/broadcom/lsp/domain/cobol/event/model/DoneAnalysisEvent.java:26]: I was inspired by the `RunAnalysisEvent`. In fact, I'm also don't like this name. I'll change it to `AnalysisFinishedEvent`.

File: com.ca.lsp.cobol/lsp-core-domain/src/main/java/com/broadcom/lsp/domain/cobol/event/model/DoneAnalysisEvent.java [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+ */
+
+package com.broadcom.lsp.domain.cobol.event.model;
+
+import lombok.Builder;
+import lombok.Data;
+import lombok.NoArgsConstructor;
+
+/**
+ * This class is a data transfer object that contains the uri of analyzed document.
```
[the remark this answers is not in the document]
Comment (grianbrcom) [reply com.ca.lsp.cobol/lsp-core-domain/src/main/java/com/broadcom/lsp/domain/cobol/event/model/DoneAnalysisEvent.java:22]: Make sense.

File: com.ca.lsp.cobol/lsp-core-domain/src/main/java/com/broadcom/lsp/domain/cobol/event/model/DoneAnalysisEvent.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (asatklichov) [reply com.ca.lsp.cobol/lsp-core-domain/src/main/java/com/broadcom/lsp/domain/cobol/event/model/DoneAnalysisEvent.java:26]: AnalysisFinishedEvent --> AnalysisCompletionEvent we use Completion term in lsp

[the remark this answers is not in the document]
Comment (temanbrcom) [reply com.ca.lsp.cobol/lsp-core-domain/src/main/java/com/broadcom/lsp/domain/cobol/event/model/DoneAnalysisEvent.java:26]: Completion in LSP is only for the autocomplete feature

File: clients/cobol-lsp-vscode-extension/src/__tests__/MiddlewareTest.ts (hunk 1 above)
Comment (grianbrcom) [reply clients/cobol-lsp-vscode-extension/src/__tests__/MiddlewareTest.ts:40]: It could be even a const. Good point.

**Pull request merged**